feat: added a UnivariateMonomial representation to reduce field ops in protogalaxy+sumcheck

zac-williamson opened the pull request:
Summary:

`client_ivc_bench.sh` benchmark has been improved by approx 10% (26218ms vs 29306ms)

In both protogalaxy + sumcheck, the basic representation of the edge of the boolean hypercube is now a degree-1 monomial instead of a MAX_RELATION_DEGREE-degree monomial

The class UnivariateMonomial can efficiently evaluate low-degree monomial relations of up to degree-2. The relations in the `relations` directory have been reworked to perform initial low-degree algebraic computations using UnivariateMonomial, only converting to a full Monomial object once the UnivariateMonomial would otherwise exceed degree-2

Reason why we do all of this:

1. for MegaFlavor, `extend_edges` was converting every flavour polynomial into a degree-11 Univariate. This was introducing 9 Fp additions * NUM_ALL_ENTITIES per row in the circuit. Given the sparse trace structure we are working with, this is a lot of computation that this PR makes redundant
2. for each relation, we check if it can be skipped by typically calling `is_zero` on a selector. The selector poly is in Univariate form (MegaFlavor = degree-11) which is 11 Fp zero-checks. MegaFlavor has 9 skippable relations which is 99 Fp zero-checks. With the new degree-2 representation this is reduced to only 18 Fp zero-checks
3. The number of raw Fp add and mul operations required to evaluate our relations is reduced. For example, in the permutation argument each `*`/`+` operation in the `accumulate` function was costing us 11 Fp muls/adds. It is cheaper to compute low-degree sub-terms in the coefficient representation before extend inginto point-evaluation representation

e.g. consider (in the protogalaxy case where challenges are degree-1 univariates) `(w_i + \beta * S_i + \gamma)` for `i = 0,1,2,3`. In coefficient representation this term can be computed with 8 Fp adds and 3 Fp muls. Extending into a degree-11 point evaluation form costs 18 Fp adds for a total of 26 Fp adds and 3 Fp muls.

In master branch, using Univariate<11> this computation costs us 20  Fp adds and 10 Fp muls. Assuming an add is 1/3 the cost of a mul, this makes the new approach cost 35 Fp add-equivalent operations vs 50 Fp add-equivalent

Overall in the new approach, the number of field operations to compute the permutation argument has reduced by 30%

zac-williamson committed:
add monomial accumulator

zac-williamson committed:
wip

zac-williamson committed:
relation updates

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
wip

zac-williamson committed:
Merge branch 'master' into zw/monomial-accumulaator

zac-williamson committed:
small cleanup

zac-williamson committed:
test inlines

zac-williamson committed:
add short univariate case to sumcheck prover

zac-williamson committed:
move small computations within thread block

zac-williamson committed:
tweaks

zac-williamson committed:
tweaks

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/generated/flavor.hpp:
```diff
@@ -121,6 +121,9 @@ class AvmFlavor {
     using RelationSeparator = AvmFlavorSettings::RelationSeparator;
 
+    // indicates when evaluating sumcheck, edges must be extended to be MAX_TOTAL_RELATION_LENGTH
+    static constexpr bool USE_SHORT_MONOMIALS = false;
```
Please port this change in bb-pilcom/bb-pil-backend/templates/flavor.hpp.hbs
Alternatively, you can ping AVM team and we can do it.

Note that the recursive flavor .hpp counterpart is not generated (therefore no change there)

zac-williamson replied:
just updated flavor.hpp.hbs

jeanmon commented on barretenberg/cpp/src/barretenberg/vm/avm/recursion/recursive_flavor.hpp:
```diff
@@ -25,6 +25,9 @@ template <typename BuilderType> class AvmRecursiveFlavor_ {
     using Relations = AvmFlavor::Relations_<FF>;
 
+    // indicates when evaluating sumcheck, edges must be extended to be MAX_TOTAL_RELATION_LENGTH
+    static constexpr bool USE_SHORT_MONOMIALS = false;
```
Would it make sense to align it with the native flavor?
What I mean is to set it:
`    static constexpr bool USE_SHORT_MONOMIALS = NativeFlavor::USE_SHORT_MONOMIALS;
`

zac-williamson replied:
good call, updated

jeanmon reviewed (commented):
I just looked at the AVM related changes.

zac-williamson committed:
updated flavor.hpp.hbs to generate USE_SHORT_MONOMIALS

zac-williamson committed:
recursive_flavor derives USE_SHORT_MONOMIALS

zac-williamson committed:
compiler fixes

maramihali commented on barretenberg/cpp/src/barretenberg/ultra_honk/decider_keys.hpp:
```diff
@@ -69,6 +69,22 @@ template <typename Flavor_, size_t NUM_ = 2> struct DeciderProvingKeys_ {
+        // Set the size corresponding to the number of rows in the execution trace
+        // Iterate over the prover polynomials' views corresponding to each proving key
+        for (size_t dpk_idx = 0; auto& get_all : prover_polynomials_views) {
+            // Iterate over all columns in the trace execution of an proving key and extract their value at row_idx.
```
```suggestion
            // Iterate over all columns in the trace execution of a proving key and extract their value at row_idx.
```

zac-williamson replied:
fixed

maramihali commented on barretenberg/cpp/src/barretenberg/ultra_honk/decider_keys.hpp:
```diff
@@ -69,6 +69,22 @@ template <typename Flavor_, size_t NUM_ = 2> struct DeciderProvingKeys_ {
+        typename Flavor::template ProverUnivariates<2> results;
+        // Set the size corresponding to the number of rows in the execution trace
+        // Iterate over the prover polynomials' views corresponding to each proving key
+        for (size_t dpk_idx = 0; auto& get_all : prover_polynomials_views) {
```
get_all can be made `const` I believe and made renamed `view`

zac-williamson replied:
could you clarify by what you mean by "renamed view"? I don't fully understand

maramihali replied:
sorry "made renamed" was a typo, I meant rename to `view` and make const. But no strong preference to the name, was just thinking that you're iteration over something called views

maramihali commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck.hpp:
```diff
@@ -115,7 +115,8 @@ transcript. These operations are taken care of by \ref bb::BaseTranscript "Trans
  */
 template <typename Flavor> class SumcheckProver {
-
+    // PartiallyEvaluatedMultivariates OR ProverPolynomials
```
maybe move this comment above ProverPolynomials and PartiallyEvaluatedMultivariates

zac-williamson replied:
✅

maramihali commented on barretenberg/cpp/src/barretenberg/protogalaxy/protogalaxy_prover_internal.hpp:
```diff
@@ -43,6 +43,10 @@ template <class DeciderProvingKeys_> class ProtogalaxyProverInternal {
     using ExtendedUnivariateWithRandomization =
         Univariate<FF, (Flavor::MAX_TOTAL_RELATION_LENGTH - 1 + DeciderPKs::NUM - 1) * (DeciderPKs::NUM - 1) + 1>;
+
+    // TODO: describe wtf is going on here
```
is it clarified? :))

zac-williamson replied:
I have attempted to now :)

maramihali commented on barretenberg/cpp/src/barretenberg/protogalaxy/protogalaxy_prover_internal.hpp:
```diff
@@ -292,18 +296,28 @@ template <class DeciderProvingKeys_> class ProtogalaxyProverInternal {
-    static void extend_univariates(
-        std::conditional_t<skip_count != 0, ExtendedUnivariates, ExtendedUnivariatesNoOptimisticSkipping>&
+    BB_INLINE static void extend_univariates(
+        std::conditional_t<
```
think it would help readability to define this conditional (that now has an inner conditional) separately

maramihali commented on barretenberg/cpp/src/barretenberg/protogalaxy/protogalaxy_prover_internal.hpp:
```diff
@@ -292,18 +296,28 @@ template <class DeciderProvingKeys_> class ProtogalaxyProverInternal {
+        if constexpr (Flavor::USE_SHORT_MONOMIALS) {
+            extended_univariates = std::move(keys.row_to_short_univariates(row_idx));
+        } else {
+            constexpr size_t _skip_count = Flavor::USE_SHORT_MONOMIALS ? 0 : skip_count;
```
is this still necessary? given the folding flavor (Mega) is now using short monomials by default, doesn't that mean we're never skipping values in this function and the `skip_count` logic should be removed? Maybe I'm misunderstanding something?

zac-williamson replied:
you are correct. I kept this code in in case we ever wanted to fold a flavor that was not amenable to this monomial abstraction. For example the AVM, ECCVM or Translator circuits. Wdyt should we keep this code path in or remove?

maramihali replied:
I think I would personally incline for removing, we can always bring it back from git history, this code is a little monster already.

maramihali commented on barretenberg/cpp/src/barretenberg/protogalaxy/protogalaxy_prover_internal.hpp:
```diff
@@ -391,42 +405,43 @@ template <class DeciderProvingKeys_> class ProtogalaxyProverInternal {
         using ExtendedUnivatiatesType =
-            std::conditional_t<skip_zero_computations, ExtendedUnivariates, ExtendedUnivariatesNoOptimisticSkipping>;
+
+            std::conditional_t<Flavor::USE_SHORT_MONOMIALS,
```
same question here as to whether we still do optimistic skipping

zac-williamson replied:
I think that we do still do optimistic skipping. I tried to trace where this type is used, and it seems to only be used when initialising the `AllEntities` data structure passed into the `accumulate` method for our relations.

In our relation classes, the `Accumulator` type is derived from the types in `ContainerOverSubrelations`. This in turn is the same as the type `TupleOfTuplesOfUnivariates_` param in `accumulate_relation_univariates` from `protogalaxy_prover_impl.hpp`, which is the same as the `TupleOfTuples` param in `compute_combiner` which is, from what I can tell, the `TupleOfTuplesOfUnivariates` type defined in `protogalaxy_prover.hpp` and this type *does* support skipping

Phew. I might try to rationalise this a bit as I am not sure the template parameter propagation in the protogalaxy methods is required as the type used is defined in the class

zac-williamson replied:
I just pushed a change which removes the template parameter types in `compute_combiner` and `accumulate_relation_univariates`. The only time we need to not optimitically skip seems to be in `combiner.test.cpp` (which is why we needed parametrisable types), so I created bespoke methods for `combiner.test.cpp` by defining a new class `PGInternalTest` that inherits from `ProtogalaxyProverInternal` and adds back in methods that enable not skipping.

While this is more lines of code overall, I think it makes `ProtogalaxyProverInternal` cleaner and easier to read, as there are fewer template parameter propagation chains to follow when examining the code. wdyt?

maramihali commented on barretenberg/cpp/src/barretenberg/relations/auxiliary_relation.hpp:
```diff
@@ -155,55 +159,56 @@ template <typename FF_> class AuxiliaryRelationImpl {
+        limb_subproduct += (w_1_shift_m * w_2_shift_m);
+        auto non_native_field_gate_1_m = limb_subproduct;
+        non_native_field_gate_1_m -= (w_3_m + w_4_m);
+        auto non_native_field_gate_1 = ShortAccumulator(non_native_field_gate_1_m) * ShortAccumulator(q_3_m);
```
will you add a brief comment as to why the `ShortAccumulator` transformation is required here?

zac-williamson replied:
added comment

maramihali commented on barretenberg/cpp/src/barretenberg/relations/databus_lookup_relation.hpp:
```diff
@@ -184,16 +185,18 @@ template <typename FF_> class DatabusLookupRelationImpl {
     static Accumulator compute_write_term(const AllEntities& in, const Parameters& params)
     {
+        using MonomialAccumulator = typename Accumulator::MonomialAccumulator;
         using View = typename Accumulator::View;
```
View and ParameterView seem to be not used anymore?

zac-williamson replied:
fixed

maramihali commented on barretenberg/cpp/src/barretenberg/relations/ecc_op_queue_relation.hpp:
```diff
@@ -67,17 +67,17 @@ template <typename FF_> class EccOpQueueRelationImpl {
-        auto lagrange_ecc_op = View(in.lagrange_ecc_op);
+        using MonomialAccumulator = typename Accumulator::MonomialAccumulator;
+
+        auto w_1 = Accumulator(MonomialAccumulator(in.w_l));
```
why do we handle things differently here in comparison to other relations?

zac-williamson replied:
I will add a comment explaining this

maramihali commented on barretenberg/cpp/src/barretenberg/relations/poseidon2_internal_relation.hpp:
```diff
@@ -62,53 +62,61 @@ template <typename FF_> class Poseidon2InternalRelationImpl {
+        auto u4_m = MonomialAccumulator(in.w_4);
 
+        auto q_pos_by_scaling_m = (q_poseidon2_internal_m * scaling_factor);
+        auto q_pos_by_scaling = Accumulator(q_pos_by_scaling_m);
```
The call to `Accumulator` is because we need extra values to properly perform subsequent computations right?

zac-williamson replied:
correct

maramihali reviewed (commented):
Left a couple of questions and requests for  extra tests of interaction between`Univariate` and `UnivariateMonomial` classes

zac-williamson committed:
reduced tparams in ProtogalaxyProverInternal

Moved methods in ProtogalaxyProverInternal that did not optimistically skip into a separate class `PGInternalTest` in `combiner.test.cpp`

zac-williamson commented on barretenberg/cpp/src/barretenberg/relations/ultra_arithmetic_relation.hpp:
```diff
@@ -83,46 +83,98 @@ template <typename FF_> class UltraArithmeticRelationImpl {
+            // tmp *= scaling_factor;
+            // std::get<0>(evals) += tmp;
+
+            using MonomialAccumulator = typename Accumulator::MonomialAccumulator;
```
good call

zac-williamson commented on barretenberg/cpp/src/barretenberg/relations/ultra_arithmetic_relation.hpp:
```diff
@@ -83,46 +83,98 @@ template <typename FF_> class UltraArithmeticRelationImpl {
                                   const FF& scaling_factor)
     {
         PROFILE_THIS_NAME("Arithmetic::accumulate");
```
some of the later terms in the relation will be of the Accumulator type (e.g. `tmp0` in this file).

I figured as a general rule it might be easier to read what's going on by applying a heuristic that `_m` indicates the type is a MonomialAccumulator, given we use `auto` type declarations for frequently

zac-williamson commented on barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/mega_recursive_flavor.hpp:
```diff
@@ -40,6 +40,8 @@ template <typename BuilderType> class MegaRecursiveFlavor_ {
     using NativeVerificationKey = NativeFlavor::VerificationKey;
 
+    // indicates when evaluating sumcheck, edges can be left as degree-1 monomials
+    static constexpr bool USE_SHORT_MONOMIALS = true;
```
good call

zac-williamson committed:
Updated to reflect PR comments

zac-williamson commented on barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/ultra_recursive_flavor.hpp:
```diff
@@ -54,6 +54,9 @@ template <typename BuilderType> class UltraRecursiveFlavor_ {
     using NativeVerificationKey = NativeFlavor::VerificationKey;
 
+    // indicates when evaluating sumcheck, edges can be left as degree-1 monomials
+    static constexpr bool USE_SHORT_MONOMIALS = true;
```
fixed

zac-williamson committed:
more PR feedback

zac-williamson commented on barretenberg/cpp/src/barretenberg/eccvm/eccvm_flavor.hpp:
```diff
@@ -40,6 +40,9 @@ class ECCVMFlavor {
     using MSM = bb::eccvm::MSM<CycleGroup>;
 
+    // indicates when evaluating sumcheck, edges must be extended to be MAX_TOTAL_RELATION_LENGTH
+    static constexpr bool USE_SHORT_MONOMIALS = false;
```
fixed. I'm curious why the tests all passed without this. Do we not compile tests that use the sumcheck Prover for the translator and eccvm recursive flavours?

maramihali replied:
oh, now that I think about it, this flag is only useful for the prover, and given we only ever instantiate the verifiers with recursive flavors the flag is not necessary in them, my bad

zac-williamson commented on barretenberg/cpp/src/barretenberg/polynomials/univariate.hpp:
```diff
@@ -47,6 +50,73 @@ template <class Fr, size_t domain_end, size_t domain_start = 0, size_t skip_coun
     Univariate& operator=(const Univariate& other) = default;
     Univariate& operator=(Univariate&& other) noexcept = default;
 
+    explicit operator UnivariateMonomial<Fr, 2, true>() const
```
good call. will add a test suite when I can find some time to do so!

zac-williamson committed:
more PR feedback

zac-williamson committed:
added tests for UnivariateMonomial

zac-williamson committed:
renamed "UnivariateMonomial" to "UnivariateCoefficientBasis"

renamed "MonomialAccumulator" to "CoefficientAccumulator"

zac-williamson commented on barretenberg/cpp/src/barretenberg/polynomials/univariate_monomial.hpp:
```diff
@@ -0,0 +1,343 @@
+    static_assert(LENGTH == 2 || LENGTH == 3);
+    using value_type = Fr; // used to get the type of the elements consistently with std::array
+
+    std::array<Fr, 3> coefficients;
```
added

zac-williamson commented on barretenberg/cpp/src/barretenberg/polynomials/univariate.hpp:
```diff
@@ -47,6 +50,73 @@ template <class Fr, size_t domain_end, size_t domain_start = 0, size_t skip_coun
+        evaluations[domain_end - 1] = prev;
+    }
+
+    // explicit operator UnivariateMonomial<Fr, 2, 0, 0>() const
```
fixed

maramihali commented on barretenberg/cpp/src/barretenberg/polynomials/univariate.hpp:
```diff
@@ -102,17 +102,17 @@ template <class Fr, size_t domain_end, size_t domain_start = 0, size_t skip_coun
     }
 
-    // explicit operator UnivariateMonomial<Fr, 2, 0, 0>() const
+    // explicit operator UnivariateCoefficientBasis<Fr, 2, 0, 0>() const
```
stale?

maramihali reviewed (approved):
Thanks for the updates!

zac-williamson committed:
Merge branch 'master' into zw/monomial-accumulaator

zac-williamson committed:
PR comments

zac-williamson committed:
Merge branch 'master' into zw/monomial-accumulaator

zac-williamson committed:
Merge branch 'master' into zw/monomial-accumulaator
